### DISCUSSION TITLE
Product existence check / Asynchronous domain

### DISCUSSION THREAD

**Discussion by chrishanzlik** [kgrzybek/sample-dotnet-core-cqrs-api#31] [Q&A] [answered]
Hi Kamil,

very great example and reading by the way!

How can you make sure that the products you are adding to an order (`PlaceCustomerOrderCommand`) really exists?
It may of course be that this topic has been left out for code brevity.
But it seems still important to me to check that the `ProductDto` with the given `ProductId` really exists, so that we do not generate any data garbage.

Where do you think this check would take place? In my opinion it is a domain concern. Adding an invalid item to the invoice would break a domain rule. In this case, an `IProductExistenceChecker` could help. This would be implemented in a similar way to the `CustomerUniquessChecker` and passed as a dependency / parameter. What do you think about it?

But that leads me to another question. `CustomerUniquessChecker` is using a **synchronous** database call to check if the email already exists. Since this check does not take place very often, it may be okay in this case. But when we check the existence of a product, this query is used far more frequently.
Would it make sense to design the domain asynchronously to prevent a bottleneck? For example: 

``` csharp
public async Task<OrderId> PlaceOrderAsync(
    List<OrderProductData> orderProductsData,
    List<ProductPriceData> allProductPrices,
    string currency, 
    List<ConversionRate> conversionRates,
    IProductExistenceChecker productExistenceChecker)
{
    await CheckRuleAsync(new ProductsExistRule(orderProductsData, productExistenceChecker));
    CheckRule(new CustomerCannotOrderMoreThan2OrdersOnTheSameDayRule(_orders));
    CheckRule(new OrderMustHaveAtLeastOneProductRule(orderProductsData));

    var order = Order.CreateNew(orderProductsData, allProductPrices, currency, conversionRates);

    this._orders.Add(order);

    this.AddDomainEvent(new OrderPlacedEvent(order.Id, this.Id, order.GetValue()));

    return order.Id;
}
```

Or should caching be the way to go and stay synchronous?

**Comment by kgrzybek** [comment, marked as answer]
Hi @chrishanzlik 

From my experience, the best place to check if something exists is an Application Layer (Application Service / Command Handler). If a product does not exist it means that the command is invalid so the operation should be aborted immediately (fail-fast principle). In Application Service you can use async/await so it will be more efficient too.

However, this is the case when the boundary between the application layer and domain layer is blurry. So you can put this logic in Domain but in my opinion, it is not a good idea and I would stick to the Application Layer. I try to avoid I/O operation in Domain Layer (even hidden behind the interface) and make Domain pure (but this is only a general rule)
